instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/Instrumenter.java

```diff
@@ -222,16 +233,23 @@ private void doEnd(
       span.recordException(error);
     }
 
-    UnsafeAttributes attributes = new UnsafeAttributes();
+    State state = context.get(INSTRUMENTER_STATE_CONTEXT_KEY);
+    UnsafeAttributes attributes = state == null ? new UnsafeAttributes() : state.attributes();
+    if (state == null) {
+      logger.log(
+          Level.FINE,
+          "No instrumenter state present when ending context {0}.  Cannot run operationListeners; metrics may not be recorded",
+          context);
+    }
     for (AttributesExtractor<? super REQUEST, ? super RESPONSE> extractor : attributesExtractors) {
       extractor.onEnd(attributes, context, request, response, error);
     }
     span.setAllAttributes(attributes);
```

Comment:
I think we only want to set end attributes here on the span since users could have chosen to overwrite some of the start attributes